### DISCUSSION TITLE
ComfyUI Search and replace strings in WAS Save Text File and Save Image nodes

### DISCUSSION THREAD

**Discussion by masslevel** [WASasquatch/was-node-suite-comfyui#129] [Ideas]
Maybe I haven't figured out the correct syntax, but is it possible to use the Search and replace strings of ComfyUI in WAS Save nodes? This way WAS custom tokens and widget information could be used to construct filenames.

Right now the Search and replace strings feature only seems to work in the default ComfyUI Save Image node.

I tried putting `%node_name.widget_name%` into the filename_prefix widget in the WAS Image Save Node and also the Text Add Tokens node. But the variables were not populated.

The search and replace strings feature is unofficially documented here: https://blenderneko.github.io/ComfyUI-docs/Interface/SaveFileFormatting/

**Comment by WASasquatch** [comment]
As far as I knew these were part of a custom_node by Data It. If they are in base ComfyUI it's a bug that they don't work on custom nodes.

**Reply by masslevel** [reply]
Thanks for your reply and thank you for your work, @WASasquatch! The WAS node suite helped me to build some really cool components for my workflow.

I see.  I didn't know that. I thought because the `Node name for S&R` option is available in every native and custom node under the node's properties, it would be something custom nodes could hook into.

![image](https://github.com/WASasquatch/was-node-suite-comfyui/assets/2313001/29e605d6-8c72-4a20-94d5-9e2c96e78902)

**Reply by WASasquatch** [reply]
It's front-end functionality, so there wouldn't even by anything for custom_nodes to do, it would be replaced by the front-end JavaScript handling the *widgets*. "widgets" don't exist on the back end. That's why I suggested it may be a bug.

**Comment by WinstonWoof** [comment]
May I turn this on its head a little then.

The image Below shows 4 WAS Nodes

If I wanted to pass the  Checkpoint name to a text file output how would I do it?

![image](https://github.com/WASasquatch/was-node-suite-comfyui/assets/128506111/8b393261-17a8-4101-a5d3-8079c46afe2f)

**Comment by WASasquatch** [comment]
Convert text_b to a input and hook up the checkpoint name output from the
loader to it.

**Reply by WinstonWoof** [reply]
will give it a whirl

**Reply by WinstonWoof** [reply]
Ok so that works but then it also shows the issue.

If I want to , for example, pull in the Sampler Name then because even your version of the Ksampler doesnt have the text outputs I cant connect it the same way.

*IF* your Nodes were able to use the built in S&R Functionality this would be far easier for the end user IMHO
![image](https://github.com/WASasquatch/was-node-suite-comfyui/assets/128506111/fffda82e-f1b6-4d25-8a13-89b0400754e8)

**Comment by WASasquatch** [comment]
You would need to bring that up with ComfyUI. That isn't a my nodes issues.
That's ComfyUI and using LiteGraph which is a mess.

**Reply by WinstonWoof** [reply]
so using you Nodes only  how would I pass the sampler name & scheduler name to the .txt output please?

**Comment by WASasquatch** [comment]
Currently you wouldn't until ComfyUI fixes that and allows widget tokens to
be used in custom node fields I guess.

Even the primitive node is handled on the front end so not sure I could
make a node that converts the combo value to a string.

The only work around I can think of which just dawned on me before sending
this, is to print the widget code in primitive under util. That in theory
should print the value in it, which then could be used as the value of a
WAS-NS token which can be printed in the WAS-NS save node.

**Reply by WinstonWoof** [reply]
see I was thinking of it the other way.

Something similar to your Parse Tokens node that could Parse S&R terms and pipe them to a text output

**Comment by WASasquatch** [comment]
Problem with that, again, is it would be a custom node and thus S&R
wouldn't work. I'd have the implement it myself, and I don't cover for bugs
in mainline. They should just be fixed.
